Src/main.c

```diff
@@ -100,14 +107,36 @@ int main(void)
   debug("\n\nBootcamp starting up...");
   debug("Compiled on %s at %s", __DATE__, __TIME__);
 
-
   HAL_GPIO_WritePin(LED1_GPIO_Port, LED1_Pin, GPIO_PIN_SET);
+
+  HAL_ADC_Start(&hadc); //Enables ADC and starts conversion of the regular channels.
+  HAL_TIM_PWM_Start(&htim16, TIM_CHANNEL_1); //starts PWM signal generation. First argument is the pointer to the TIM handle and the second parameter is the channel used or the timer
+
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
+
+  struct bit_field b; //defines bitfield which will contain the 12 bit adc_output
+  b.adc_output = 0; 
+
+  //int adc_output = 0; //ADC gives a 12 bit output 
+
+  uint16_t pwm_compare = 0; //we only need to provide a 16 bit input for the pwm
+
   while (1)
   {
+    b.adc_output = HAL_ADC_GetValue(&hadc); //Gets the converted value from data register of regular channel. (A value between 0 and 4095 representinng 0 to 3.3 V [assuming a 12-bit ADC is being used])
+    /*
+    
+      From my understanding, the __HAL_TIM_SET_COMPARE() function compares the pwm_compare value to the period of the TIM16. The value a = pwm_compare/TIM16.Period*100% is the duty period of the PWM.
+      Since we want 1ms to 2ms on-time, we are looking at a 5-10% duty cycle. Right now, the PWM period is 60000, thus we want the pwm_compare value to have a range of 3000 to 6000!
+
+    */
+    pwm_compare = b.adc_output * 3000 / 4095 + 3000; // Converts value to a range of 3000 (1 ms on-time) to 6000 (2 ms on-time)
```

Comment:
Another thing: it's good practice to avoid random numbers when you're dealing with peripherals, as it requires someone to go into the datasheet, figure out how you came up with those numbers, and then reverse-engineer it. An example could be: what if we asked for you to use a servo with a different minimum and maximum duty cycle specification? Comments are nice, but it's much easier to just change a constant at the top than to redo the math :)

Reply:
Good point!